### ISSUE TITLE
AWS S3 artifact repo: Failed to put file: 404 Not Found

### ISSUE THREAD

**Issue by barambani** [argoproj/argo#1784]
**Is this a BUG REPORT or FEATURE REQUEST?**:
Unexpected behaviour (possibly a bug ?)

**What happened**:
After deploying argo to EKS (AWS) through argo CD and configuring the default artifact repository to use S3 directly, a simple example workflow (see below) fails to Put the compressed object to the bucket with a message `404 Not Found`. I use the IAM role (as can be seen from the logs). Same behaviour if I use a non default repo in the Workflow template's definition.

*The relevant step pod's `wait` container logs*:
```
[...]
time="2019-11-20T10:26:51Z" level=info msg="Saving output artifacts"
time="2019-11-20T10:26:51Z" level=info msg="Annotations monitor stopped"
time="2019-11-20T10:26:51Z" level=info msg="Staging artifact: licence-output-artifact"
time="2019-11-20T10:26:51Z" level=info msg="Copying /tmp/licence.txt from container base image layer to /tmp/argo/outputs/artifacts/licence-output-artifact.tgz"
time="2019-11-20T10:26:51Z" level=info msg="Archiving a578930e494911a3d7488c17f7244416cc55f7404c0d7fb92ff9b5b27210690d:/tmp/licence.txt to /tmp/argo/outputs/artifacts/licence-output-artifact.tgz"
time="2019-11-20T10:26:51Z" level=info msg="sh -c docker cp -a a578930e494911a3d7488c17f7244416cc55f7404c0d7fb92ff9b5b27210690d:/tmp/licence.txt - | gzip > /tmp/argo/outputs/artifacts/licence-output-artifact.tgz"
time="2019-11-20T10:26:51Z" level=info msg="Archiving completed"
time="2019-11-20T10:26:51Z" level=info msg="S3 Save path: /tmp/argo/outputs/artifacts/licence-output-artifact.tgz, key: licence-output-artifact.tgz"
time="2019-11-20T10:26:51Z" level=info msg="Creating minio client s3.amazonaws.com using IAM role"
time="2019-11-20T10:26:51Z" level=info msg="Saving from /tmp/argo/outputs/artifacts/licence-output-artifact.tgz to s3 (endpoint: s3.amazonaws.com, bucket: argo-workflow-s3-artifact-repo-dev-us-east-1, key: licence-output-artifact.tgz)"
time="2019-11-20T10:26:51Z" level=warning msg="Failed to put file: 404 Not Found"
time="2019-11-20T10:26:52Z" level=info msg="Deadline monitor stopped"
[...]
```

**What you expected to happen**:
I was expecting to see the artifact stored in the bucket.

**How to reproduce it (as minimally and precisely as possible)**:
On an Argo EKS deployment these are my settings

*My repo config map*:
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: workflow-controller-configmap
data:
  config: |
    artifactRepository:
      s3:
        bucket: argo-workflow-s3-artifact-repo-dev-us-east-1
        keyPrefix: test
        endpoint: s3.amazonaws.com
```

*My test workflow*:
```yaml
spec:
  arguments:
    parameters:
    - name: url
      value: https://raw.githubusercontent.com/EarnestResearch/aws-lambda-haskell-runtime-client/master/LICENSE
  entrypoint: http-sequential-pipeline
  serviceAccountName: pipeline-runner
  templates:
  - name: http-download
    inputs:
      parameters:
      - name: url
    container:
      args:
      - '{{inputs.parameters.url}}'
      command:
      - sh
      - -c
      - 'curl {{inputs.parameters.url}} > /tmp/licence.txt'
      image: byrnedo/alpine-curl:0.1.8
    outputs:
      artifacts:
      - name: licence-output-artifact
        path: /tmp/licence.txt
        s3:
          endpoint: s3.amazonaws.com
          bucket: argo-workflow-s3-artifact-repo-dev-us-east-1
          key: licence-output-artifact.tgz
  - name: echo
    inputs:
      artifacts:
      - name: licence-input-artifact
        path: /tmp/licence.txt
    container:
      args:
      - "echo /tmp/licence.txt"
      command:
      - sh
      - -c
      image: library/bash:4.4.23
  - dag:
      tasks:
      - name: http-download
        template: http-download
        arguments:
          parameters:
          - name: url
            value: '{{inputs.parameters.url}}'
      - name: echo
        template: echo
        arguments:
          artifacts:
          - name: licence-input-artifact
            from: '{{dag.tasks.http-download.outputs.artifacts.licence-output-artifact}}'
        dependencies:
        - http-download
    inputs:
      parameters:
      - name: url
    name: http-sequential-pipeline
```

**Anything else we need to know?**:
I have tried the following things:
- I have given List Object and Write Object permission to everyone during one test assuming a valid (verified) role explicitly using
```yaml
outputs:
      artifacts:
      - name: licence-output-artifact
        path: /tmp/licence.txt
        s3:
          endpoint: s3.amazonaws.com
          bucket: argo-workflow-s3-artifact-repo-dev-us-east-1
          key: licence-output-artifact.tgz
          roleARN: arn:aws:iam::<an alloewd role>
```
but nothing changed.
- I've verified that the bucket is reachable from inside the kube cluster adding manually a pod with aws CLI and pushing files to the bucket from the same namespace as the argo step pods.

Side note: There is a recurring message in the workflow controller's logs that I'm not sure what mean. Something like.
```
[...]
W1120 10:56:38.905526       1 reflector.go:289] github.com/argoproj/argo/workflow/controller/config_controller.go:116: watch of *v1.ConfigMap ended with: too old resource version: 839943 (841602)
time="2019-11-20T10:58:46Z" level=info msg="Alloc=5167 TotalAlloc=69307 Sys=70590 NumGC=538 Goroutines=69"
[...]
```

**Environment**:
- Argo version:
```
2.4.1
```
- Kubernetes version :
```
$ kubectl version -o yaml
clientVersion:
  buildDate: "2019-11-14T04:24:34Z"
  compiler: gc
  gitCommit: b3cbbae08ec52a7fc73d334838e18d17e8512749
  gitTreeState: clean
  gitVersion: v1.16.3
  goVersion: go1.12.13
  major: "1"
  minor: "16"
  platform: darwin/amd64
serverVersion:
  buildDate: "2019-10-18T17:56:01Z"
  compiler: gc
  gitCommit: b7174db5ee0e30c94a0b9899c20ac980c0850fc8
  gitTreeState: clean
  gitVersion: v1.14.8-eks-b7174d
  goVersion: go1.12.10
  major: "1"
  minor: 14+
  platform: linux/amd64
```

**Other debugging information (if applicable)**:
- workflow result:
```
$ argo get <workflowname>
```
- executor logs:
see above

- workflow-controller logs:
```
$ kubectl logs -n argo $(kubectl get pods -l app=workflow-controller -n argo -o name)
[...]
time="2019-11-20T10:18:46Z" level=info msg="Alloc=5217 TotalAlloc=65416 Sys=70590 NumGC=518 Goroutines=69"
time="2019-11-20T10:23:46Z" level=info msg="Alloc=5143 TotalAlloc=65471 Sys=70590 NumGC=520 Goroutines=69"
time="2019-11-20T10:26:49Z" level=info msg="Processing workflow" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:49Z" level=info msg="Updated phase  -> Running" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:49Z" level=info msg="DAG node http-sequential-pipeline-jdd24 (http-sequential-pipeline-jdd24) initialized Pending" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:49Z" level=info msg="node http-sequential-pipeline-jdd24 (http-sequential-pipeline-jdd24) phase Pending -> Running" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:49Z" level=info msg="All of node http-sequential-pipeline-jdd24.http-download dependencies [] completed" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:49Z" level=info msg="Pod node http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740) initialized Pending" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:49Z" level=info msg="Created pod: http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740)" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:49Z" level=info msg="Workflow update successful" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:50Z" level=info msg="Processing workflow" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:50Z" level=info msg="Updating node http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740) message: ContainerCreating"
time="2019-11-20T10:26:50Z" level=info msg="Skipped pod http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740) creation: already exists" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:50Z" level=info msg="Workflow update successful" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:51Z" level=info msg="Processing workflow" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:51Z" level=info msg="Updating node http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740) status Pending -> Running"
time="2019-11-20T10:26:51Z" level=info msg="Skipped pod http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740) creation: already exists" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:51Z" level=info msg="Workflow update successful" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:52Z" level=info msg="Processing workflow" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:26:52Z" level=info msg="Skipped pod http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740) creation: already exists" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:23Z" level=info msg="Processing workflow" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:23Z" level=info msg="Setting node http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740) outputs"
time="2019-11-20T10:27:23Z" level=info msg="Skipped pod http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740) creation: already exists" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:23Z" level=info msg="Workflow update successful" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:24Z" level=info msg="Processing workflow" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:24Z" level=info msg="Updating node http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740) status Running -> Succeeded"
time="2019-11-20T10:27:24Z" level=info msg="Skipped node http-sequential-pipeline-jdd24.echo (http-sequential-pipeline-jdd24-2752390000) initialized Error (message: Unable to resolve: {{dag.tasks.http-download.outputs.artifacts.licence-output-artifact}})" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:24Z" level=info msg="DAG outbound nodes of http-sequential-pipeline-jdd24.http-download (http-sequential-pipeline-jdd24-3293193740) are [http-sequential-pipeline-jdd24-3293193740]" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:24Z" level=info msg="node http-sequential-pipeline-jdd24 (http-sequential-pipeline-jdd24) phase Running -> Error" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:24Z" level=info msg="node http-sequential-pipeline-jdd24 (http-sequential-pipeline-jdd24) finished: 2019-11-20 10:27:24.823098191 +0000 UTC" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:24Z" level=info msg="Checking daemoned children of http-sequential-pipeline-jdd24" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:24Z" level=info msg="Updated phase Running -> Error" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:24Z" level=info msg="Marking workflow completed" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:24Z" level=info msg="Checking daemoned children of " namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:24Z" level=info msg="Workflow update successful" namespace=kubeflow workflow=http-sequential-pipeline-jdd24
time="2019-11-20T10:27:25Z" level=info msg="Labeled pod kubeflow/http-sequential-pipeline-jdd24-3293193740 completed"
W1120 10:27:29.888462       1 reflector.go:289] github.com/argoproj/argo/workflow/controller/config_controller.go:116: watch of *v1.ConfigMap ended with: too old resource version: 830333 (832546)
time="2019-11-20T10:28:46Z" level=info msg="Alloc=5253 TotalAlloc=68130 Sys=70590 NumGC=523 Goroutines=69"
W1120 10:33:16.892010       1 reflector.go:289] github.com/argoproj/argo/workflow/controller/config_controller.go:116: watch of *v1.ConfigMap ended with: too old resource version: 833125 (834375)
time="2019-11-20T10:33:46Z" level=info msg="Alloc=5324 TotalAlloc=68275 Sys=70590 NumGC=525 Goroutines=69"
time="2019-11-20T10:38:46Z" level=info msg="Alloc=5165 TotalAlloc=68586 Sys=70590 NumGC=528 Goroutines=69"
W1120 10:42:40.897741       1 reflector.go:289] github.com/argoproj/argo/workflow/controller/config_controller.go:116: watch of *v1.ConfigMap ended with: too old resource version: 834942 (837280)
time="2019-11-20T10:43:46Z" level=info msg="Alloc=5241 TotalAlloc=68718 Sys=70590 NumGC=530 Goroutines=69"
time="2019-11-20T10:48:46Z" level=info msg="Alloc=5187 TotalAlloc=68793 Sys=70590 NumGC=533 Goroutines=69"
W1120 10:49:16.901618       1 reflector.go:289] github.com/argoproj/argo/workflow/controller/config_controller.go:116: watch of *v1.ConfigMap ended with: too old resource version: 837859 (839333)
time="2019-11-20T10:53:46Z" level=info msg="Alloc=5185 TotalAlloc=68886 Sys=70590 NumGC=535 Goroutines=69"
W1120 10:56:38.905526       1 reflector.go:289] github.com/argoproj/argo/workflow/controller/config_controller.go:116: watch of *v1.ConfigMap ended with: too old resource version: 839943 (841602)
time="2019-11-20T10:58:46Z" level=info msg="Alloc=5167 TotalAlloc=69307 Sys=70590 NumGC=538 Goroutines=69"
```

**Comment by sarabala1979** [comment]
We will try to reproduce. It should work. Basically, Argo is using `minio client` push and retrieve the object from S3. The error looks like a Container is not able to connect `s3.amazon.com` or error might misleading with some other error.  Can you try to add `region` and try it?

**Comment by barambani** [comment]
Thanks, I have tried adding the region with the same outcome. It looks like that the `Not Found` error was actually hiding an role issue. I could make it work adding a role annotation to every step of the DAG. Something like
```yaml
metadata:
      annotations:
        iam.amazonaws.com/role: <a role allowed to S3>
```
I still see a couple of problems with the use case above. First of all the error message. `404 Not Found` is misleading if the problem is permission or credentials related. The other one is that when using a non-default repo, `roleARN` didn't make it work, so it seems it's not effective ? Not sure.

To give more context I'm using Kiam in the cluster and what I'd like to do is to give the right annotation to all the pods created by the workflow controller so that they can access S3. Is there a nicer way than forcing all the step definitions to have it ? Thanks a lot for the responsiveness.

**Comment by alexec** [comment]
I think the 404 may come from the API. Often APIs return 404 rather than 401/403 so they do not leak information.

**Comment by candonov** [comment]
I am running into the exact same issue. (Running argo v2.8.0-rc4) with IRSA (as well as Kiam) for auth and neither is working. I started digging into the code I noticed argo is using v1.27.1 of `aws-skd-go` https://github.com/argoproj/argo/blob/48d86f03ce9ac49f5099c46dd7995305eb59eb11/go.mod#L12 which is about 6 months old. I've ran into issues with IRSA before using `aws cli` and the solution was to upgrade to the latest version. I am not further enough in this investigation to be able to confirm that theory.
